docs/guides/advanced/testing.md

```diff
@@ -0,0 +1,183 @@
+      nextPageNumber: 2,
+    });
+    const page2Response = newJsonFetchResponse({
+      users: [{ id: 456, name: 'Bob' }],
+      nextPageNumber: undefined,
+    });
+    syncContext.fetcher.fetch
+      .withArgs('/api/users')
+      .returns(page1Response)
+      .withArgs('/api/users?page=2')
+      .returns(page2Response);
+
+    const result =
+        await executeSyncFormulaFromPackDef(pack, 'MySync', [], syncContext);
+
+    assert.equal(result.length, 2);
+    assert.equal(result[0].Id, 123);
+    assert.equal(result[1].Id, 456);
+    sinon.assert.calledTwice(syncContext.fetcher.fetch);
+  });
+});
+```
+
+### Integration test {: #integration}
+
+If you wish to write an end-to-end integration test that actually hits the third-party API that you Pack interacts with, you can simply pass `useRealFetcher: true` when using these test utilities. The execution context will include a fetcher that will make real HTTP requests to whatever urls they are given. For example:
+
+```ts
+import {executeFormulaFromPackDef} from '@codahq/packs-sdk/dist/development';
+import {pack} from '../pack';
+
+describe('Formula integration test', () => {
+  it('executes the formula', async () => {
+    const result = await executeFormulaFromPackDef(
+      pack,
+      "MyFormula",
+      ["my-param"],
+      undefined,
+      undefined,
+      {
+        useRealFetcher: true,
+        manifestPath: require.resolve("../pack"),
+      },
+    );
+    assert.equal(result, 'my-return-value');
+  });
+});
+```
+
+The fetcher will apply authentication to these requests if you have configured authentication locally using `coda auth`. For this to work you must specify the `manifestPath` and set it to the directory where the `.coda-credentials.json` file is located (usually the same directory as the Pack definition).
+
+
+### Return value validation
+
+By default, these testing utility functions will validate return values after executing your Pack formulas. This validation checks that the values you actually return from your formula implementations match the schema you have written. This helps find bugs in your code and also helps catch subtle issues in how your values might be interpreted in the Coda application when you Pack is executed for real.
+
+This validation can also help ensure that your test code correctly simulates responses from the API that you’re integrating with. For instance, while developing our Pack, you may have been regularly exercising your formula code by running `coda execute --fetch` frequently and you’re confident that your code works correctly when run against the real API. Then you go to write unit tests for you Pack and you define some fake response objects, but you forget some required fields or you specified a field as an array when it should be a comma-separated list. If your fake response result in your Pack is returning a value that doesn’t match the schema you defined, the validator will catch these and notify you.
```

Comment:
Same, can probably remove `--fetch` now that it's the default (well, after we release the next npm version)

Reply:
Done